runtime/compiler/optimizer/OSRGuardInsertion.cpp

```diff
@@ -30,63 +30,205 @@
+         {
+         TR::Block* block = treeTop->getEnclosingBlock();
+         protectedByOSRPoints = !guardAnalysis || guardAnalysis->_blockAnalysisInfo[block->getNumber()]->isEmpty();
+         continue;
+         }
+
+      TR::Node *osrNode = NULL;
+      if (comp()->isPotentialOSRPoint(ttNode, &osrNode))
+         {
+         if (visited.contains(osrNode))
+            continue;
+
+         if (protectedByOSRPoints &&
+             osrNode->isPotentialOSRPointHelperCall())
+            {
+            if (trace())
+               traceMsg(comp(), "Remove redundant potentialOSRPointHelper call n%dn %p\n", osrNode->getGlobalIndex(), osrNode);
+
+            TR::TreeTop* prevTree = treeTop->getPrevTreeTop();
+            TR::TransformUtil::removeTree(comp(), treeTop);
+            treeTop = prevTree;
+            }
+         else if (comp()->isPotentialOSRPointWithSupport(treeTop))
+            {
+            if (!protectedByOSRPoints && trace())
+               traceMsg(comp(), "treetop n%dn is an OSR point with support\n", ttNode->getGlobalIndex());
+
+            protectedByOSRPoints = true;
+            }
+         else
+            {
+            if (protectedByOSRPoints && trace())
+               traceMsg(comp(), "treetop n%dn is an OSR point without support\n", ttNode->getGlobalIndex());
+
+            protectedByOSRPoints = false;
+            }
+
+         visited.add(osrNode);
+         continue;
+         }
       }
-   if (guard == virtualGuards.end())
+   }
+
+static bool skipOSRGuardInsertion(TR::Compilation* comp)
+   {
+   // Use with caution. There can be fear points generated before OSRGuardInsertion,
+   // disabling it will leave those fear points unprotected.
+   //
+   static char *disableOSRGuards = feGetEnv("TR_DisableOSRGuards");
+
+   // Currently, OSR guard insertion is only needed when in the OSR HCR mode
+   if (disableOSRGuards || !comp->isOSRTransitionTarget(TR::postExecutionOSR))
       {
-      if (trace())
-         traceMsg(comp(), "No HCR guards to be removed - skipping\n");
-      return 0;
-      } 
+      return true;
+      }
 
-   TR_BitVector fearGeneratingNodes(comp()->getNodeCount(), trMemory(), stackAlloc);
-   removeHCRGuards(fearGeneratingNodes);
-   // Future fear generating optimizations
+   // Even under NextGenHCR, OSR may have been disabled for this compilation at runtime
+   if (!comp->supportsInduceOSR())
+      {
+      return true;
+      }
 
-   if (fearGeneratingNodes.isEmpty())
+   return false;
+   }
+
+int32_t TR_OSRGuardInsertion::perform()
+   {
+   bool needHCRGuardRemoval = hasHCRGuard(comp());
+   bool canInsertOSRGuards = !skipOSRGuardInsertion(comp());
+
+   if (canInsertOSRGuards && needHCRGuardRemoval)
       {
-      if (trace())
-         traceMsg(comp(), "No fear generating nodes - skipping\n");
-      comp()->getFlowGraph()->invalidateStructure();
-      return 0;
+      bool hasBadYieldPoint = hasYieldPointCannotOSR(comp());
```

Comment:
hasBadYieldPoint should be something like hasUnsupportedOSRPoint or something like that...